### PR TITLE
Add fallback if driver not supports isValid

### LINKED ISSUE #1504: HikariCp is not able to connect to DB2 Database
When i am trying to connect to DB2 using using Spring Boot 2.2.1. HikariCP Pool  is not able to make a successful connection but at the same time it is connecting to JDBC.

Adding clue:
When i tried to put wrong password it gave me authentication error.

Enviroments:

 1. HikariCP version: 3.4.1
 2. JDK version     : 1.8.0_111
 3. Database        : DB2
 4. Driver version  : 11.5
 5. Spring boot Version : 2.2.1

Below are error logs

        2019-11-28 16:16:38.098  INFO 10144 --- [  restartedMain] o.s.b.w.embedded.tomcat.TomcatWebServer  : Tomcat initialized with port(s): 8080 (http)
    2019-11-28 16:16:38.196  INFO 10144 --- [  restartedMain] o.apache.catalina.core.StandardService   : Starting service [Tomcat]
    2019-11-28 16:16:38.197  INFO 10144 --- [  restartedMain] org.apache.catalina.core.StandardEngine  : Starting Servlet engine: [Apache Tomcat/9.0.27]
    2019-11-28 16:16:38.799  INFO 10144 --- [  restartedMain] o.a.c.c.C.[Tomcat].[localhost].[/]       : Initializing Spring embedded WebApplicationContext
    2019-11-28 16:16:38.799  INFO 10144 --- [  restartedMain] o.s.web.context.ContextLoader            : Root WebApplicationContext: initialization completed in 8427 ms
    2019-11-28 16:16:39.866  WARN 10144 --- [  restartedMain] com.zaxxer.hikari.HikariConfig           : HikariPool-1 - maxLifetime is less than 30000ms, setting to default 1800000ms.
    2019-11-28 16:16:39.866  INFO 10144 --- [  restartedMain] com.zaxxer.hikari.HikariDataSource       : HikariPool-1 - Starting...
    2019-11-28 16:16:42.035  INFO 10144 --- [  restartedMain] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Driver does not support get/set network timeout for connections. ([jcc][10181][13991][4.26.14] Method getNetworkTimeout is not supported. ERRORCODE=-4450, SQLSTATE=0A504)
    2019-11-28 16:16:42.036 ERROR 10144 --- [  restartedMain] com.zaxxer.hikari.pool.PoolBase          : HikariPool-1 - Failed to execute isValid() for connection, configure connection test query ([jcc][1079][11955][4.26.14] Feature not supported: timeout in isValid is not supported. ERRORCODE=-4450, SQLSTATE=0A504).
    2019-11-28 16:16:42.320 ERROR 10144 --- [  restartedMain] com.zaxxer.hikari.pool.HikariPool        : HikariPool-1 - Error thrown while acquiring connection from data source
    
    com.ibm.db2.jcc.am.SqlFeatureNotSupportedException: [jcc][1079][11955][4.26.14] Feature not supported: timeout in isValid is not supported. ERRORCODE=-4450, SQLSTATE=0A504
    	at com.ibm.db2.jcc.am.b7.a(b7.java:804) ~[jcc-11.5.0.0.jar:na]
    	at com.ibm.db2.jcc.am.b7.a(b7.java:66) ~[jcc-11.5.0.0.jar:na]
    	at com.ibm.db2.jcc.am.b7.a(b7.java:116) ~[jcc-11.5.0.0.jar:na]
    	at com.ibm.db2.jcc.am.Connection.isValidX(Connection.java:4158) ~[jcc-11.5.0.0.jar:na]
    	at com.ibm.db2.jcc.am.Connection.isValid(Connection.java:7702) ~[jcc-11.5.0.0.jar:na]
    	at com.zaxxer.hikari.pool.PoolBase.checkValidationSupport(PoolBase.java:450) ~[HikariCP-3.4.1.jar:na]
    	at com.zaxxer.hikari.pool.PoolBase.checkDriverSupport(PoolBase.java:433) ~[HikariCP-3.4.1.jar:na]
    	at com.zaxxer.hikari.pool.PoolBase.setupConnection(PoolBase.java:402) ~[HikariCP-3.4.1.jar:na]
    	at com.zaxxer.hikari.pool.PoolBase.newConnection(PoolBase.java:355) ~[HikariCP-3.4.1.jar:na]
    	at com.zaxxer.hikari.pool.PoolBase.newPoolEntry(PoolBase.java:201) ~[HikariCP-3.4.1.jar:na]
    	at com.zaxxer.hikari.pool.HikariPool.createPoolEntry(HikariPool.java:473) [HikariCP-3.4.1.jar:na]
    	at com.zaxxer.hikari.pool.HikariPool.checkFailFast(HikariPool.java:562) [HikariCP-3.4.1.jar:na]
    	at com.zaxxer.hikari.pool.HikariPool.<init>(HikariPool.java:115) [HikariCP-3.4.1.jar:na]
    	at com.zaxxer.hikari.HikariDataSource.getConnection(HikariDataSource.java:112) [HikariCP-3.4.1.jar:na]
    	at org.springframework.jdbc.datasource.DataSourceUtils.fetchConnection(DataSourceUtils.java:158) [spring-jdbc-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.jdbc.datasource.DataSourceUtils.doGetConnection(DataSourceUtils.java:116) [spring-jdbc-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.jdbc.datasource.DataSourceUtils.getConnection(DataSourceUtils.java:79) [spring-jdbc-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.jdbc.core.JdbcTemplate.execute(JdbcTemplate.java:324) [spring-jdbc-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.boot.jdbc.EmbeddedDatabaseConnection.isEmbedded(EmbeddedDatabaseConnection.java:120) [spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.autoconfigure.orm.jpa.HibernateDefaultDdlAutoProvider.getDefaultDdlAuto(HibernateDefaultDdlAutoProvider.java:42) [spring-boot-autoconfigure-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaConfiguration.lambda$getVendorProperties$1(HibernateJpaConfiguration.java:130) [spring-boot-autoconfigure-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.autoconfigure.orm.jpa.HibernateSettings.getDdlAuto(HibernateSettings.java:41) ~[spring-boot-autoconfigure-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.autoconfigure.orm.jpa.HibernateProperties.determineDdlAuto(HibernateProperties.java:136) ~[spring-boot-autoconfigure-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.autoconfigure.orm.jpa.HibernateProperties.getAdditionalProperties(HibernateProperties.java:102) ~[spring-boot-autoconfigure-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.autoconfigure.orm.jpa.HibernateProperties.determineHibernateProperties(HibernateProperties.java:94) ~[spring-boot-autoconfigure-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.autoconfigure.orm.jpa.HibernateJpaConfiguration.getVendorProperties(HibernateJpaConfiguration.java:132) [spring-boot-autoconfigure-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.autoconfigure.orm.jpa.JpaBaseConfiguration.entityManagerFactory(JpaBaseConfiguration.java:133) ~[spring-boot-autoconfigure-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_221]
    	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:1.8.0_221]
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:1.8.0_221]
    	at java.lang.reflect.Method.invoke(Unknown Source) ~[na:1.8.0_221]
    	at org.springframework.beans.factory.support.SimpleInstantiationStrategy.instantiate(SimpleInstantiationStrategy.java:154) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.ConstructorResolver.instantiate(ConstructorResolver.java:640) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.ConstructorResolver.instantiateUsingFactoryMethod(ConstructorResolver.java:625) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.instantiateUsingFactoryMethod(AbstractAutowireCapableBeanFactory.java:1338) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBeanInstance(AbstractAutowireCapableBeanFactory.java:1177) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.doCreateBean(AbstractAutowireCapableBeanFactory.java:557) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.AbstractAutowireCapableBeanFactory.createBean(AbstractAutowireCapableBeanFactory.java:517) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.AbstractBeanFactory.lambda$doGetBean$0(AbstractBeanFactory.java:323) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.DefaultSingletonBeanRegistry.getSingleton(DefaultSingletonBeanRegistry.java:222) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.AbstractBeanFactory.doGetBean(AbstractBeanFactory.java:321) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.beans.factory.support.AbstractBeanFactory.getBean(AbstractBeanFactory.java:202) ~[spring-beans-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.context.support.AbstractApplicationContext.getBean(AbstractApplicationContext.java:1108) ~[spring-context-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.context.support.AbstractApplicationContext.finishBeanFactoryInitialization(AbstractApplicationContext.java:868) ~[spring-context-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.context.support.AbstractApplicationContext.refresh(AbstractApplicationContext.java:550) ~[spring-context-5.2.1.RELEASE.jar:5.2.1.RELEASE]
    	at org.springframework.boot.web.servlet.context.ServletWebServerApplicationContext.refresh(ServletWebServerApplicationContext.java:141) ~[spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.SpringApplication.refresh(SpringApplication.java:747) ~[spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.SpringApplication.refreshContext(SpringApplication.java:397) ~[spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.SpringApplication.run(SpringApplication.java:315) ~[spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1226) ~[spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at org.springframework.boot.SpringApplication.run(SpringApplication.java:1215) ~[spring-boot-2.2.1.RELEASE.jar:2.2.1.RELEASE]
    	at com.alight.onelight.benefitService.BenefitServiceApplication.main(BenefitServiceApplication.java:10) ~[main/:na]
    	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method) ~[na:1.8.0_221]
    	at sun.reflect.NativeMethodAccessorImpl.invoke(Unknown Source) ~[na:1.8.0_221]
    	at sun.reflect.DelegatingMethodAccessorImpl.invoke(Unknown Source) ~[na:1.8.0_221]
    	at java.lang.reflect.Method.invoke(Unknown Source) ~[na:1.8.0_221]
    	at org.springframework.boot.devtools.restart.RestartLauncher.run(RestartLauncher.java:49) ~[spring-boot-devtools-2.2.1.RELEASE.jar:2.2.1.RELEASE]

Properties files attached:

    spring.datasource.url=jdbc:db2:DBT1
    spring.datasource.username=XXXXX
    spring.datasource.password=XXXXXXX
    spring.datasource.driver-class-name=com.ibm.db2.jcc.DB2Driver
    spring.jpa.show-sql=true
    spring.jpa.properties.hibernate.dialect=org.hibernate.dialect.DB2Dialect
    spring.datasource.hikari.minimum-idle= 10
    spring.datasource.hikari.maximum-pool-size= 10
    spring.datasource.hikari.max-lifetime= 1000
    spring.datasource.hikari.auto-commit =true

### PR DESCRIPTION
**Pull request by quaff**
Fix https://github.com/brettwooldridge/HikariCP/issues/1504

### CODE REVIEW

File: src/main/java/com/zaxxer/hikari/pool/PoolBase.java [hunk 1]
```diff
@@ -447,7 +448,12 @@ private void checkValidationSupport(final Connection connection) throws SQLExcep
    {
       try {
          if (isUseJdbc4Validation) {
```
Comment (brettwooldridge) [inline src/main/java/com/zaxxer/hikari/pool/PoolBase.java]: How does ``isUseJdbc4Validation`` get set to ``true`` if it is not supported? I would have expected ``SQLFeatureNotSupportedException`` to have been thrown earlier, forcing ``isUseJdbc4Validation=false``.

Comment (quaff) [reply src/main/java/com/zaxxer/hikari/pool/PoolBase.java]: @brettwooldridge Good point, I've revised it.

**Comment by brettwooldridge** [comment]
@quaff Wait, wait, wait. What exactly is wrong with the original code? I explicitly did *not* want a "fallback" to something like ``SELECT 1``, which is a query that does not work on many databases (HSQLDB, Oracle, etc.). In fact, [there is no validation query that works universally](https://stackoverflow.com/questions/3668506/efficient-sql-test-query-or-validation-query-that-will-work-across-all-or-most) across databases.

The intent of the existing code is:
 * If no validation query is set, then use ``isValid()``, ie. ``isUseJdbc4Validation = config.getConnectionTestQuery() == null``
 * And, if ``isValid()`` is not supported by the driver, then an exception (``ConnectionSetupException``) will be thrown and logged

This is intended to *force* a user who is using a driver without ``isValid()`` support to explicitly specify a validation query, because only the user can supply a correct one. "Falling back" to ``SELECT 1`` could lead to some very confused users when the ``SELECT 1`` query itself fails against their database and they know they did not specify any query.

Sometimes the solution to a failure due to ``isValid()`` not being supported is simply a newer JDBC driver, rather than a validation query, which is a terribly inefficient way to validate a connection and should be avoided when possible.

**Comment by quaff** [comment]
> @quaff Wait, wait, wait. What exactly is wrong with the original code? I explicitly did _not_ want a "fallback" to something like `SELECT 1`, which is a query that does not work on many databases (HSQLDB, Oracle, etc.). In fact, [there is no validation query that works universally](https://stackoverflow.com/questions/3668506/efficient-sql-test-query-or-validation-query-that-will-work-across-all-or-most) across databases.
> 
> The intent of the existing code is:
> 
> * If no validation query is set, then use `isValid()`, ie. `isUseJdbc4Validation = config.getConnectionTestQuery() == null`
> * And, if `isValid()` is not supported by the driver, then an exception (`ConnectionSetupException`) will be thrown and logged
> 
> This is intended to _force_ a user who is using a driver without `isValid()` support to explicitly specify a validation query, because only the user can supply a correct one. "Falling back" to `SELECT 1` could lead to some very confused users when the `SELECT 1` query itself fails against their database and they know they did not specify any query.
> 
> Sometimes the solution to a failure due to `isValid()` not being supported is simply a newer JDBC driver, rather than a validation query, which is a terribly inefficient way to validate a connection and should be avoided when possible.

I think this change is better than original, It doesn't change behavior of existing code, the intent is:
* `isValid` is supported by the driver, everything is OK
* `isValid` is not supported by the driver but user provide their own validation query, everything is OK
* `isValid` is not supported by the driver and user didn't provide their own validation query, if database supports `select 1` everything is OK, otherwise exception is thrown, I add a warn message that `select 1` will be used, maybe the message should be more detailed, WDYT @brettwooldridge ?

**Comment by quaff** [comment]
Added an error message to alert user provide their own connection test query if default `select 1` is not supported.

**Comment by brettwooldridge** [comment]
@quaff Ok, if we're going to go this far, then let's go further.  What I would like is a solution that is *specific* to drivers with no known JDBC4 ``isValid()`` support. For example, DB2 in your case.

Here is what I am thinking.  Add a static method to the ``UtilityElf`` class, along the direction of:
```java
private static final Map<String, String> package2queryMap;

static
{
   package2queryMap = new HashMap<>();
   package2queryMap.put("com.ibm.db2.jcc", "SELECT 1");
   package2queryMap.put("COM.ibm.db2os390.sqlj.jdbc", "SELECT 1");
}

public static String getDefaultValidationQuery(final Connection connection)
{
   final String connectionPackage = connection.getClass().getPackageName();
   return package2queryMap.get(connectionPackage);
}
```
In ``PoolBase`` you're going to need a new member variable, ``connectionTestQuery``, that is set in the constructor to ``config.getConnectionTestQuery()``.  Then in ``PoolBase.checkValidationSupport()``, if the call to ``isValid`` throws ``SQLFeatureNotSupportedException`` or ``AbstractMethodError``, catch and call the above method. If that method returns non-null then set ``connectionTestQuery`` to that query string and log a warning. If that method returns null, then simply re-throw the exception. Obviously, ``connectionTestQuery`` must be used wherever ``config.getConnectionTestQuery()`` was previously used. (I do not want to modify the configuration itself).

Bascially, over time we can "whitelist" certain known drivers with default queries. If you actually know of other specific drivers that do not support ``isValid``, feel free to add the appropriate test query. But what I *don't* want to do is to default to a query like ``SELECT 1`` that is *unknown* as to whether it is correct for the database or not.

**Comment by quaff** [comment]
> @quaff Ok, if we're going to go this far, then let's go further. What I would like is a solution that is _specific_ to drivers with no known JDBC4 `isValid()` support. For example, DB2 in your case.
> 
> Here is what I am thinking. Add a static method to the `UtilityElf` class, along the direction of:
> 
> ```java
> private static final Map<String, String> package2queryMap;
> 
> static
> {
>    package2queryMap = new HashMap<>();
>    package2queryMap.put("com.ibm.db2.jcc", "SELECT 1");
>    package2queryMap.put("COM.ibm.db2os390.sqlj.jdbc", "SELECT 1");
> }
> 
> public static String getDefaultValidationQuery(final Connection connection)
> {
>    final String connectionPackage = connection.getClass().getPackageName();
>    return package2queryMap.get(connectionPackage);
> }
> ```
> 
> In `PoolBase` you're going to need a new member variable, `connectionTestQuery`, that is set in the constructor to `config.getConnectionTestQuery()`. Then in `PoolBase.checkValidationSupport()`, if the call to `isValid` throws `SQLFeatureNotSupportedException` or `AbstractMethodError`, catch and call the above method. If that method returns non-null then set `connectionTestQuery` to that query string and log a warning. If that method returns null, then simply re-throw the exception. Obviously, `connectionTestQuery` must be used wherever `config.getConnectionTestQuery()` was previously used. (I do not want to modify the configuration itself).
> 
> Bascially, over time we can "whitelist" certain known drivers with default queries. If you actually know of other specific drivers that do not support `isValid`, feel free to add the appropriate test query. But what I _don't_ want to do is to default to a query like `SELECT 1` that is _unknown_ as to whether it is correct for the database or not.

I think this is overengineering, if `select 1` is not supported by database, there is a warning and an error message, it's enough for troubleshooting.
```
c.z.h.p.PoolBase WARN HikariPool-1 - Use default connection test query (select 1)
c.z.h.p.PoolBase ERROR HikariPool-1 - Please config your own connection test query since default one is not supported
```

**Comment by brettwooldridge** [comment]
Thanks. I'll have to pass on this pull request.

**Comment by quaff** [comment]
> I'll have to pass on this pull request

Sorry, I'm not native english speaker, Does `pass on` means approve or skip?

**Pull request closed without merging**